django/shortcuts.py

```diff
@@ -41,11 +43,19 @@ def redirect(to, *args, permanent=False, **kwargs):
         * A URL, which will be used as-is for the redirect location.
 
     Issues a temporary redirect by default; pass permanent=True to issue a
-    permanent redirect.
+    permanent redirect. Pass preserve_method=True to issue a redirect that must
+    preserve HTTP verb.
     """
-    redirect_class = (
-        HttpResponsePermanentRedirect if permanent else HttpResponseRedirect
-    )
+
+    redirect_class = HttpResponseRedirect
+
+    if not preserve_method and permanent:
+        redirect_class = HttpResponsePermanentRedirect
+    if preserve_method and not permanent:
+        redirect_class = HttpResponseTemporaryRedirectWithSameMethod
+    if preserve_method and permanent:
+        redirect_class = HttpResponsePermanentRedirectWithSameMethod
```

Comment:
Any thoughts about changing `HttpResponseRedirectBase` to accept a `preserve_method: bool = False` kwarg instead and make it assign a different status code to `self` over distinct classes?

Something like 

```diff
diff --git a/django/http/response.py b/django/http/response.py
index abe71718f2..9e701cf16c 100644
--- a/django/http/response.py
+++ b/django/http/response.py
@@ -615,7 +615,7 @@ def set_headers(self, filelike):
 class HttpResponseRedirectBase(HttpResponse):
     allowed_schemes = ["http", "https", "ftp"]

-    def __init__(self, redirect_to, *args, **kwargs):
+    def __init__(self, redirect_to, preserve_method=False, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self["Location"] = iri_to_uri(redirect_to)
         parsed = urlsplit(str(redirect_to))
@@ -623,6 +623,8 @@ def __init__(self, redirect_to, *args, **kwargs):
             raise DisallowedRedirect(
                 "Unsafe redirect to URL with protocol '%s'" % parsed.scheme
             )
+        if preserve_method:
+            self.status_code = self.preserve_method_status_code

     url = property(lambda self: self["Location"])

@@ -640,10 +642,12 @@ def __repr__(self):

 class HttpResponseRedirect(HttpResponseRedirectBase):
     status_code = 302
+    preserve_method_status_code = 307


 class HttpResponsePermanentRedirect(HttpResponseRedirectBase):
     status_code = 301
+    preserve_method_status_code = 308


 class HttpResponseNotModified(HttpResponse):
```

Reply:
Also fine with me, should I go this path instead?

Reply:
However, on a second look, it would break (in a way) the consistency of the file, where we have response classes with a single status code.

Reply:
I think the suggestion from @charettes makes sense. I would prefer using that approach to avoid defining two new classes, as naming them correctly would be challenging.

Reply:
This PR would also need to add relevant tests to `HttpResponseSubclassesTests` (in `tests/httpwrappers/tests.py`).